scripts/speech_recognition/estimate_duration_bins_2d.py

```diff
@@ -0,0 +1,319 @@
+    num_tokens = joint.f1
+
+    size_per_bucket = sizes.sum() / num_buckets
+
+    if not quiet:
+        print("Duration distribution:")
+        print(pd.Series(sizes).describe(percentiles=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]))
+    if math.isinf(max_duration):
+        max_duration = sizes[-1]
+
+    tps = num_tokens / sizes
+    if not quiet:
+        print("Token per second distribution:")
+        print(pd.Series(tps).describe(percentiles=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]))
+    if math.isinf(max_tps):
+        max_tps = tps.max()
+    del tps
+
+    bins = []
+    bin_indexes = [0]
+    tot = 0.0
+    for binidx, size in enumerate(sizes):
+        if tot > size_per_bucket:
+            num_tokens_bucket = num_tokens[bin_indexes[-1] : binidx]
+            num_tokens_bucket.sort()
+            tokens_per_subbucket = num_tokens_bucket.sum() / num_subbuckets
+            tot_toks = 0
+            for num_toks in num_tokens_bucket:
+                if tot_toks > tokens_per_subbucket:
+                    bins.append((size, num_toks))
+                    tot_toks = 0
+                tot_toks += num_toks
+            bins.append((size, math.ceil(size * max_tps)))
+            bin_indexes.append(binidx)
+            tot = 0.0
+        tot += size
+
+    num_tokens_bucket = num_tokens[bin_indexes[-1] : binidx]
+    num_tokens_bucket.sort()
+    tokens_per_subbucket = num_tokens_bucket.sum() / num_subbuckets
+    tot_toks = 0
+    for num_toks in num_tokens_bucket:
+        if tot_toks > tokens_per_subbucket:
+            bins.append((max_duration, num_toks))
+            tot_toks = 0
+        tot_toks += num_toks
+    bins.append((max_duration, math.ceil(max_duration * max_tps)))
+
+    return bins
+
+
+def load_tokenizer(paths: list[str], langs: list[str] = None) -> TokenizerWrapper:
+    if len(paths) == 1:
+        tok = SentencePieceTokenizer(paths[0])
+    else:
+        assert langs is not None and len(paths) == len(
+            langs
+        ), f"Cannot create AggregateTokenizer; each tokenizer must have assigned a language via --langs option (we got --tokenizers={paths} and --langs={langs})"
+        tok = AggregateTokenizer({lang: SentencePieceTokenizer(p) for lang, p in zip(langs, paths)})
+    return TokenizerWrapper(tok)
+
+
+def apply_tokenizer(cut, tokenizer=None, prompt: PromptFormatter = None):
+    if prompt is not None:
+        turns = prompt.get_default_dialog_slots()
+        last_turn = {"role": prompt.OUTPUT_ROLE, "slots": prompt.get_slots(prompt.OUTPUT_ROLE)}
+        assert len(last_turn["slots"]) == 1  # TODO: not sure how to handle multi-slot for system output here
+        for key in last_turn["slots"]:
+            last_turn["slots"][key] = cut.supervisions[0].text
+        last_turn["slots"][prompt.PROMPT_LANGUAGE_SLOT] = cut.supervisions[0].language
+        turns.append(last_turn)
+        ans = prompt.encode_dialog(turns)
+        cut.supervisions[0].tokens = ans["input_ids"]
```

Comment:
Just to make sure I understand, `ans["input_ids"]` contains the entire token sequence that the model will train on to predict, not just the initial prompt (i.e., " <|startoftranscript|>" and whatnot), right?

Reply:
Correct:  `input_ids == context_ids + answer_ids` (context=prompt, answer=model prediction)